crates/ika-core/src/dwallet_mpc/mpc_event.rs

```diff
@@ -0,0 +1,418 @@
+    DWalletEncryptionKeyReconfigurationRequestEvent, DWalletImportedKeyVerificationRequestEvent,
+    DWalletMPCEvent, DWalletNetworkDKGEncryptionKeyRequestEvent, DWalletSessionEvent,
+    DWalletSessionEventTrait, EncryptedShareVerificationRequestEvent, FutureSignRequestEvent,
+    MakeDWalletUserSecretKeySharesPublicRequestEvent, PresignRequestEvent, SignRequestEvent,
+    DWALLET_MODULE_NAME,
+};
+use std::time::Duration;
+use tokio::sync::broadcast;
+use tracing::{debug, error, info, warn};
+
+impl DWalletMPCManager {
+    /// Handle a batch of MPC events.
+    ///
+    /// This function might be called more than once for a given session, as we periodically
+    /// check for uncompleted events - in which case the event will be ignored.
+    ///
+    /// A new MPC session is only created once at the first time the event was received
+    /// (per-epoch, if it was uncompleted in the previous epoch it will be created again for the next one.)
+    ///
+    /// If the event already exists in `self.mpc_sessions`, we do not add it.
+    ///
+    /// If there is no session info, and we've got it in this call,
+    /// we update that field in the open session.
+    pub(crate) fn handle_dwallet_db_events(
+        &mut self,
+        events: Vec<DBSuiEvent>,
+        epoch_store: &AuthorityPerEpochStore,
+    ) {
+        // We only update `next_active_committee` here, so once we update it there is no longer going to be any pending events for it in this epoch.
+        if self.next_active_committee.is_none() {
+            let got_next_active_committee = self.try_receiving_next_active_committee();
+            if got_next_active_committee {
+                // `..` stands for `RangeFull`, and calling `drain(..)` will drain (i.e. mutate) the vector thus removing all events from it.
+                for event in self
+                    .events_pending_for_next_active_committee
+                    .drain(..)
+                    .collect::<Vec<_>>()
+                {
+                    self.handle_mpc_event(event);
+                }
+            }
+        }
+
+        // First try to update the network keys.
+        let newly_updated_network_keys_ids = self.update_network_keys();
+
+        // Then check if we just got the public data for some network keys, and handle those events if so.
+        for key_id in newly_updated_network_keys_ids {
+            for event in self
+                .events_pending_for_network_key
+                .remove(&key_id)
+                .unwrap_or_default()
+            {
+                self.handle_mpc_event(event);
+            }
+        }
+
+        for event in events {
+            self.handle_sui_event(event, epoch_store);
+        }
+    }
+
+    fn handle_sui_event(&mut self, event: DBSuiEvent, epoch_store: &AuthorityPerEpochStore) {
+        if event.type_.address != *epoch_store.packages_config.ika_system_package_id
```

Comment:
This kind of check was implicit later in the code (when comparing the type of the event with the packages config), but its an exteremly important security check and it must be highlighted and explicit. Otherwise, later someone might remove some code that wasn't commented to have security effects, and a vulnerability could be inserted unknowingly.